src/replication/master_replication_channel_manager.hpp

```diff
@@ -0,0 +1,161 @@
+	if (m_master_daemon_threads[index] != NULL)
+	  {
+	    cubthread::get_manager()->destroy_daemon_without_entry (m_master_daemon_threads[index]);
+	  }
+	ptr->set_channel (this->m_channel);
+	/* TODO[arnia] in the future maybe useful to have more control over looper */
+	m_master_daemon_threads[index] = cubthread::get_manager ()->create_daemon_without_entry (cubthread::looper (
+					   std::chrono::milliseconds (10)),
+					 ptr, master_replication_channel::master_daemon_names[index]);
+      }
+
+      ~master_replication_channel_entry ();
+
+      std::shared_ptr<master_replication_channel> &get_replication_channel();
+
+      master_replication_channel_entry (master_replication_channel_entry &&entry);
+      master_replication_channel_entry &operator= (master_replication_channel_entry &&entry);
+
+      master_replication_channel_entry (const master_replication_channel_entry &entry) = delete;
+      master_replication_channel_entry &operator= (const master_replication_channel_entry &entry) = delete;
+  };
+
+  class master_replication_channel_manager
+  {
+    public:
+      static void init (cubstream::stream *stream);
+      static void add_master_replication_channel (master_replication_channel_entry &&channel);
+      static unsigned int get_number_of_channels ();
+      static void reset ();
+
+      static cubstream::stream_position g_minimum_successful_stream_position;
+
+      static inline cubstream::stream &get_stream ()
+      {
+	assert (g_stream != NULL);
+	return *g_stream;
+      }
+
+    private:
+      class master_channels_supervisor_task : public cubthread::entry_task
+      {
+	public:
+	  master_channels_supervisor_task ()
+	  {
+	  }
+
+	  void execute (cubthread::entry &context)
+	  {
+	    static unsigned int check_conn_delay_counter = 0;
+
+	    if (check_conn_delay_counter >
+		SUPERVISOR_DAEMON_CHECK_CONN_MS / SUPERVISOR_DAEMON_DELAY_MS)
+	      {
+		auto new_end = std::remove_if (master_channels.begin (), master_channels.end (),
+					       [] (master_replication_channel_entry &entry)
+		{
+		  return !entry.get_replication_channel ()->is_connected ();
+		});
+		master_channels.erase (new_end, master_channels.end ());
+
+		check_conn_delay_counter = 0;
+	      }
+
+	    master_replication_channel_manager::g_minimum_successful_stream_position =
+	      std::numeric_limits <cubstream::stream_position>::max();
+
+	    for (auto &entry : master_channels)
+	      {
+		if (master_replication_channel_manager::g_minimum_successful_stream_position >
+		    entry.get_replication_channel ()->get_current_sending_position ())
+		  {
+		    master_replication_channel_manager::g_minimum_successful_stream_position =
+		      entry.get_replication_channel ()->get_current_sending_position ();
+		  }
+	      }
+
+	    check_conn_delay_counter++;
+	  }
+      };
+
+      master_replication_channel_manager ();
+      ~master_replication_channel_manager ();
+
+      static std::vector <master_replication_channel_entry> master_channels;
```

Comment:
consider multi-threaded access to master_channels; a synchronization mechanism may be required.

Reply:
done